---
title: MPBIO-BBT045 Course page
date: "2025"
theme: "simplex"
output:
  html_document:
    number_sections: false
    toc: true
    toc_float: true
    toc_depth: 3
    df_print: paged
  pdf_document: default
---

# Homework

* [HW 1: Unix](homework1)
* [HW 2: Python for Data Analysis](python_for_data_analysis/homework_python_intro.html)
* [HW 3: Alignment](alignment/homework.html)
* [HW 4: HMM](HMM/HMM_notebook.html)
* [HW 5: Phylogeny](https://v-varga.github.io/BBT045-Phylo-2024/Homework/PhyloHomework_py.html)

<hr />


# Study Material, Tutorials and Exercises

## Tutorial 1: Introduction to Unix

**Preliminaries**

* Official documentation for the Vera cluster at C3SE can be found [here](https://www.c3se.chalmers.se/documentation/for_users/getting_started/how_to_login_to_vera/).
* Introduction slides can be found [here](Unix_intro.pdf).

**Tutorial**

* [Introduction to Unix](unix-tasks)

**Homework**

* [HW 1: Unix](homework1)  
* To be submitted on Canvas by January 29th.

**Recommended Reading**

* The Unix Shell tutorial from the Software Carpentry Foundation: https://swcarpentry.github.io/shell-novice/
* Unix commands [cheat sheet](https://www.git-tower.com/blog/command-line-cheat-sheet/)
* [Unix philosophy](unix-philosophy)
* [Regular expression learning resources](regex)

<hr />

## Tutorial 2: Introduction to Python programming for data analysis

**Tutorial**

[Python Intro Tutorial](python_for_data_analysis/intro_python_data_analysis.html)

**Homework**

* [HW 2: Python for Data Analysis](python_for_data_analysis/homework_python_intro.html)
* To be submitted on Canvas by February 5th.

**Reading**

A selection of small chapters and sections that explain concepts we cover in the course:

- [*An Introduction to Programming for Bioscientists: A Python-Based Primer*](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004867) by Ekmekci et al.
    - This PLoS article is a good overview/introduction to the use of Python for data science. 
    - Some aspects may seem confusing in the beginning, but Python (like all languages, programming or otherwise) is best learned through _use_. As you work your way through this course, this article can be used a reference. 

- [*Python for Data Analysis, 3E*](https://wesmckinney.com/book/) by Wes McKinney
    - An alternative option for exploring the use of Python in the context of data analysis & scientific computing
    - Much more detailed than the PLoS article above, but less relevant for bioinformatics - a potentially useful reference

- [*Intro to Advanced Python*](https://python-course.eu/advanced-python/) by Bernd Klein
    - This website provides resources for some of the more advanced utility of Python, which can improve your overall coding ability. 
    - The [*Intro to Python Tutorial*](https://python-course.eu/python-tutorial/) portion of the same website has more basic introductory materials, if you like all of your references to be in the same place. 

- Official documentation pages: You do not need to read these in their entirety, but they can provide helpful information for a variety of the modules we will be using. Here, I will include a few of the basics: 
    - Python programming language official website: https://www.python.org/
    - Pandas documentation: https://pandas.pydata.org/
    - Numpy documentation: https://numpy.org/
    - Scipy (Python for scientific computing): https://scipy.org/
    - Re (regex/regular expressions in Python): https://docs.python.org/3/library/re.html
    - Biopython: https://biopython.org/

- Miscellaneous :) There are many good resource on the web. We'll point out some when relevant.

<hr />

## Tutorial 3: Version control with Git

**Preparation**  
Do this **before** you start with the tutorial.

* Step 1: [Set up a github account](Git/github_account)
* Step 2: [Set up a ssh key on the server](Git/github_SSH_key)

**Tutorial**

* [Introduction to Git](git-exercise-1)

**Recommended Reading**

* [Interactive Git cheatsheet](https://ndpsoftware.com/git-cheatsheet.html#loc=workspace;)
* [Git Cheat Sheet (PDF)](https://github.github.com/training-kit/downloads/github-git-cheat-sheet.pdf)
* [Slides](Git_intro_presentation.pdf)


<hr />

## Tutorial 4: Sequencing Technologies

[Tutorial](seq_tech/seq_tech_tutorial_py.html)

<hr />

## Tutorial 5: Introduction to Algorithmic Thinking

[Tutorial](intro_to_algos/intro_to_algos_2025.html)

In [this](https://github.com/bengtssonpalme/MPBIO-BBT045-2025/blob/master/python_benchmaking.ipynb) jupyter notebook you can find a comparison of the time needed by some of the methods you used to work with lists and Pandas data frames during the tutorial. 

<hr />

## Tutorial 6: Nextflow

[Tutorial](Nextflow_tutorial/Nextflow_tutorial.html)

<hr />

## Tutorial 7: Phylogeny

* [Tutorial](https://v-varga.github.io/BBT045-Phylo-2024/Tutorial/PhyloTutorial_py.html)
* [Homework](https://v-varga.github.io/BBT045-Phylo-2024/Homework/PhyloHomework_py.html)
* [Phylogeny website](https://v-varga.github.io/BBT045-Phylo-2024/)
* [Phylogeny GitHub Repository](https://github.com/V-Varga/BBT045-Phylo-2024/tree/main?tab=readme-ov-file)

<hr />

## Tutorial 8: Gene prediction/annotation

* [Tutorial](gene-prediction-exercise.html)

<hr />

## Tutorial: Metagenomics and Single-Cell workflows

[Tutorial](single_cell/sc_analysis.html)

<hr />


# Practical info for projects 

* [Instructions for using the project container](ProjectInfo/README.html)

* [How to install your own software on the server](installing_software_on_the_server)

<hr />

# Practical info for clusters

* [How to run Jupyter Lab](Jupyter_and_Containers/Run_Jupyter.html)

<hr />

Chalmers University of Technology

<footer style="font-size:0.8em">

<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">
<img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png" />
</a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.



</footer>
